testing/src/test/java/io/stargate/it/driver/CqlSessionExtension.java

```diff
@@ -0,0 +1,405 @@
+ * created or not, accordingly.
+ *
+ * <p>If the extension is declared both on the class level and on some methods, the method-level
+ * annotations are ignored. If you need method-level sessions in addition to the global one, it is
+ * possible to do it manually by injecting a builder (see the "Parameter injection" section below).
+ *
+ * <h3>Keyspace creation</h3>
+ *
+ * By default, the extension creates a dedicated keyspace for the lifecycle of each created session,
+ * and drops it immediately after the test class/method has executed. Keyspace names are generated
+ * in a way that guarantees uniqueness, to avoid conflicts between tests.
+ *
+ * <p>A {@code USE ks} statement is automatically issued, so that the session is already "connected"
+ * to the keyspace when your test code starts executing. In general you don't need to prefix table
+ * names in your queries, and tests are often oblivious to the actual keyspace name. However it can
+ * be injected if you need it (see the "Parameter injection" section below).
+ *
+ * <h3>Customization</h3>
+ *
+ * Certain aspects of the session creation can be controlled with the {@link CqlSessionSpec}
+ * annotation. See the javadocs of that type for more details.
+ *
+ * <h3>Parameter injection</h3>
+ *
+ * The extension will detect and inject the following parameters on JUnit test methods
+ * (constructors, BeforeXxx or test methods, etc):
+ *
+ * <ul>
+ *   <li>any parameter of type {@link CqlSession}: injected with the session that was created by the
+ *       extension.
+ *   <li>any parameter of type {@link CqlSessionBuilder}: injected with a new session builder
+ *       pre-configured to connect to the Stargate cluster. This allows you to create new sessions
+ *       with a different configuration in your test methods. Note that such sessions are not
+ *       managed by the extension: <b>you must close them explicitly once you are done</b>, to avoid
+ *       resource leaks. A try-with-resources block is generally the best way to do that. Also, you
+ *       might want to invoke {@link #awaitAllNodes(CqlSession, StargateEnvironmentInfo)} if there
+ *       are multiple Stargate nodes.
+ *   <li>a parameter of type {@link CqlIdentifier} <b>AND annotated with {@link TestKeyspace}</b>:
+ *       injected with the identifier of the keyspace that was created by the extension. This is
+ *       useful if you need to look up schema metadata, use the keyspace from another session or
+ *       API, etc.
+ * </ul>
+ */
+public class CqlSessionExtension
+    implements BeforeAllCallback,
+        BeforeEachCallback,
+        AfterEachCallback,
+        AfterAllCallback,
+        ParameterResolver {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CqlSessionExtension.class);
+  private static final int KEYSPACE_NAME_MAX_LENGTH = 48;
+
+  // Tracks whether the extension was configured per-class or per-method.
+  private volatile Lifecycle sessionLifecycle;
```

Comment:
`volatile` may be unnecessary... my understanding is that JUnit will create different contexts (and consequently extension instances) for each thread during concurrent execution.

Reply:
Did you find a definitive answer for that? Unfortunately it's not really covered in the documentation...

It should be pretty easy to test with a dummy extension.

Reply:
not definitive, but I'm pretty sure I was context creation being done after assigning test task to threads when I was digging around JUnit 5 code.... but it's worth double-checking. Having thought about this some more, I'm not sure about the case when test methods are executed by different threads.

Reply:
OK, after a few tests it turns out that extensions can be accessed concurrently.

Method-level extensions: obviously not a problem because each method gets its own extension instance.
Class-level extension, across different test classes: same.

BUT with class-level extension, across test methods in the same class: if we allow parallel method execution (`junit.jupiter.execution.parallel.mode.default = concurrent`), then the same extension instance is accessed from different threads:
```
beforeAll, extension=ThreadPrintingExtension@4ea56a23, thread=ForkJoinPool-1-worker-1
beforeEach, extension=ThreadPrintingExtension@4ea56a23, thread=ForkJoinPool-1-worker-1
beforeEach, extension=ThreadPrintingExtension@4ea56a23, thread=ForkJoinPool-1-worker-2
```
So I'm keeping the fields volatile for safe publication. Note that we don't need stronger synchronization, because beforeAll clearly happens-before beforeEach of any method.

We probably don't want parallel method execution though, because this could easily have unintended side-effects if the tests are not carefully built around it. For example two methods writing and deleting rows in the same table. But there's nothing in the extension that formally prevents it. I'll add a note in the javadocs.

Reply:
Nice. Thanks for the research @olim7t ! :)